File: src/uuidv7.java
import java.nio.ByteBuffer;
import java.security.SecureRandom;
import java.util.UUID;

public class UUIDv7 {
    private static final SecureRandom random = new SecureRandom();

    public static UUID randomUUID() {
        byte[] value = randomBytes();
        ByteBuffer buf = ByteBuffer.wrap(value);
        long high = buf.getLong();
        long low = buf.getLong();
        return new UUID(high, low);
    }

    public static byte[] randomBytes() {
        // random bytes
        byte[] value = new byte[16];
        random.nextBytes(value);

        // current timestamp in ms
        ByteBuffer timestamp = ByteBuffer.allocate(Long.BYTES);
        timestamp.putLong(System.currentTimeMillis());

        // timestamp
        System.arraycopy(timestamp.array(), 2, value, 0, 6);

        // version and variant
        value[6] = (byte) ((value[6] & 0x0F) | 0x70);
        value[8] = (byte) ((value[8] & 0x3F) | 0x80);

        return value;
    }

    public static void main(String[] args) {
        var uuid = UUIDv7.randomUUID();
        System.out.println(uuid);
    }
}

File: src/uuidv7.clj
;; UUIDv7 implementation in Clojure.
;; License: Public Domain.

(ns uuidv7
  (:import (java.security SecureRandom)
           (java.util UUID)))

(defonce ^SecureRandom rng (SecureRandom.))

(defn uuid-v7 []
  (let [ts (System/currentTimeMillis)
        ;; timestamp (48 bits)
        hi-bits (bit-or (bit-shift-left ts 16)
                        ;; version + rand_a (12 bits)
                        (bit-or 0x7000 (bit-and (.nextLong rng) 0x0FFF)))
        ;; variant (10xx) + rand_b (62 bits)
        lo-bits (bit-or (bit-shift-left 2 62)
                        (bit-and (.nextLong rng) 0x3FFFFFFFFFFFFFFF))]
    (UUID. hi-bits lo-bits)))

(println (str (uuid-v7)))
